lib/meta.nix

```diff
@@ -99,4 +99,27 @@ rec {
   availableOn = platform: pkg:
     lib.any (platformMatch platform) pkg.meta.platforms &&
     lib.all (elem: !platformMatch platform elem) (pkg.meta.badPlatforms or []);
+
+  /* Get the corresponding attribute in lib.licenses
+     from the SPDX ID.
+     For SPDX IDs, see
+     https://spdx.org/licenses
+
+     Type:
+       getLicenseFromSpdxId :: str -> AttrSet
+
+     Example:
+       lib.getLicenseFromSpdxId "MIT" == lib.licenses.mit
+       => true
+       lib.getLicenseFromSpdxId "mIt" == lib.licenses.mit
+       => true
+       lib.getLicenseFromSpdxId "MY LICENSE"
+       => { shortName = "MY LICENSE"; }
+  */
+  getLicenseFromSpdxId =
+    let
+      spdxLicenses = lib.mapAttrs (id: ls: assert lib.length ls == 1; builtins.head ls)
+        (lib.groupBy (l: lib.toLower l.spdxId) (lib.filter (l: l ? spdxId) (lib.attrValues lib.licenses)));
```

Comment:
isn't this way less efficient than the old solution but more or less equal?

Reply:
From the [comment](https://github.com/NixOS/nixpkgs/pull/141221#discussion_r727290849) by @Infinisil, the old method requires linear search in every query, while the new method caches the attribute set for future queries.

Here's the corresponding old solution:
```Nix
getLicenseFromSpdxId = licstr:
  lib.findFirst
    (l: l ? spdxId && l.spdxId == licstr)
    { shortName = licstr; }
    (builtins.attrValues lib.licenses);
```

Reply:
BTW, how do we betchmark them?

Reply:
That does caching? I thought it would reintroduce the value at every execution and the let would need to be outside but I don't know Nix that well.

Reply:
Oh, now I see. I missed the `licstr:`